subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -0,0 +1,814 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <net/nrf_cloud_codec.h>
+#include "nrf_cloud_mem.h"
+#include "nrf_cloud_codec_internal.h"
+
+LOG_MODULE_REGISTER(nrf_cloud_codec, CONFIG_NRF_CLOUD_LOG_LEVEL);
+
+static int json_decode(struct nrf_cloud_obj *const obj, const struct nrf_cloud_data *const input)
+{
+	cJSON *json;
+	int ret = -ENOMSG;
+
+	/* Only try to parse JSON if the data is null-terminated */
+	if (memchr(input->ptr, '\0', input->len + 1) != NULL) {
+		json = cJSON_Parse(input->ptr);
+		if (json) {
+			obj->json = json;
+			obj->type = NRF_CLOUD_OBJ_TYPE_JSON;
+		}
+	}
+
+	return ret;
+}
+
+static int cbor_decode(struct nrf_cloud_obj *const obj, const struct nrf_cloud_data *const input)
+{
+	/* CBOR decoding not yet supported */
+	return  -ENOTSUP;
+}
+
+int nrf_cloud_obj_input_decode(struct nrf_cloud_obj *const obj,
+	const struct nrf_cloud_data *const input)
+{
+	if (!obj || !input || !input->ptr || !input->len) {
+		return -EINVAL;
+	}
+
+	int ret;
+	bool known_type = true;
+
+	/* If no valid type is provided, try to decode all known types until one succeeds */
+	if (NRF_CLOUD_OBJ_TYPE_VALID(obj) == false) {
+		obj->type = NRF_CLOUD_OBJ_TYPE__UNDEFINED;
+		known_type = false;
+	}
+
+	if ((obj->type == NRF_CLOUD_OBJ_TYPE_JSON) || !known_type) {
+		ret = json_decode(obj, input);
+		if ((ret == 0) || known_type) {
+			return ret;
+		}
+	}
+
+	if ((obj->type == NRF_CLOUD_OBJ_TYPE_CBOR) || !known_type) {
+		ret = cbor_decode(obj, input);
+		if ((ret == 0) || known_type) {
+			return ret;
+		}
+	}
+
+	return -ENOMSG;
+}
+
+int nrf_cloud_obj_msg_check(const struct nrf_cloud_obj *const obj, const char *const app_id,
+	const char *const msg_type)
+{
+	if (!obj || !obj->json || (!app_id && !msg_type)) {
+		return -EINVAL;
+	}
+
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		char *str = NULL;
+		bool match = true;
+
+		if (!obj->json) {
+			return -ENOENT;
+		}
+
+		if (app_id) {
+			(void)get_string_from_obj(obj->json, NRF_CLOUD_JSON_APPID_KEY, &str);
+			match &= (str && (strcmp(str, app_id) == 0));
+		}
+
+		if (msg_type) {
+			str = NULL;
+			(void)get_string_from_obj(obj->json, NRF_CLOUD_JSON_MSG_TYPE_KEY, &str);
+			match &= (str && (strcmp(str, msg_type) == 0));
+		}
+
+		return match ? 0 : -ENOMSG;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_str_get(const struct nrf_cloud_obj *const obj, const char *const key,
+	char **str)
+{
+	if (!obj || !key || !str) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		return get_string_from_obj(obj->json, key, str);
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_msg_init(struct nrf_cloud_obj *const obj, const char *const app_id,
```

Comment:
If we have the zcbor section allocate an array of N zcbor_state_t, then this function will need a parameter to set N which would be unused for JSON.

Reply:
how about handling the N value as follows:
```
struct nrf_cloud_cbor {
	uint_fast32_t states;
	/* TODO:
	struct zcbor_state_t *zcb;
	*/
};

/** @brief Object used for building nRF Cloud messages. */
struct nrf_cloud_obj {

	enum nrf_cloud_obj_type type;
	union {
		cJSON *json;
		struct nrf_cloud_cbor cbor;
	};

	/** Source of encoded data */
	enum nrf_cloud_enc_src enc_src;
	/** Encoded data */
	struct nrf_cloud_data encoded_data;
};

/** @brief Define an nRF Cloud CBOR object.
 *
 * This macro defines a codec object with the type of NRF_CLOUD_OBJ_TYPE_CBOR.
 *
 * @param _name	Name of the object.
 */
#define NRF_CLOUD_OBJ_CBOR_DEFINE(_name, _states) \
struct nrf_cloud_obj _name = { .type = NRF_CLOUD_OBJ_TYPE_CBOR, \
				.enc_src = NRF_CLOUD_ENC_SRC_NONE, \
				.encoded_data = { .ptr = NULL, .len = 0 } \
				.cbor = { .states = _states } }

```

Reply:
In this scenario, who allocates and frees the `states` count of `struct zcbor_state_t`?

Reply:
it would be allocated in `nrf_cloud_obj_init ` and freed in `nrf_cloud_obj_free`

Reply:
OK, sounds good to me.